client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,789 @@
+  });
+};
+
+const runAccessibilityChecks = async (
+  onClickSelector: (selectorName: string, event: MouseEvent) => void,
+) => {
+  const a11yRowTemplate = document.querySelector<HTMLTemplateElement>(
+    '#w-a11y-result-row-template',
+  );
+  const checksPanel = document.querySelector<HTMLElement>(
+    '[data-checks-panel]',
+  );
+  const config = getAxeConfiguration(document.body);
+  const toggleCounter = document.querySelector<HTMLElement>(
+    '[data-side-panel-toggle="checks"] [data-side-panel-toggle-counter]',
+  );
+  const panelCounter = document.querySelector<HTMLElement>(
+    '[data-side-panel="checks"] [data-a11y-result-count]',
+  );
+
+  if (
+    !checksPanel ||
+    !a11yRowTemplate ||
+    !config ||
+    !toggleCounter ||
+    !panelCounter
+  ) {
+    return;
+  }
+
+  // Ensure we only test within the preview iframe, but nonetheless with the correct selectors.
+  config.context = {
+    include: {
+      fromFrames: ['#w-preview-iframe'].concat(
+        (config.context as ContextObject).include as string[],
+      ),
+    },
+  } as ContextObject;
+  if ((config.context.exclude as string[])?.length > 0) {
+    config.context.exclude = {
+      fromFrames: ['#w-preview-iframe'].concat(
+        config.context.exclude as string[],
+      ),
+    } as ContextObject['exclude'];
+  }
+
+  const { results, a11yErrorsNumber } = await getA11yReport(config);
+
+  toggleCounter.innerText = a11yErrorsNumber.toString();
+  toggleCounter.hidden = a11yErrorsNumber === 0;
+  panelCounter.innerText = a11yErrorsNumber.toString();
+  panelCounter.classList.toggle('has-errors', a11yErrorsNumber > 0);
+
+  renderA11yResults(
+    checksPanel,
+    results,
+    config,
+    a11yRowTemplate,
+    onClickSelector,
+  );
+};
+
+interface PreviewDataResponse {
+  is_valid: boolean;
+  is_available: boolean;
+}
+
+/**
+ * Controls the preview panel component to submit the current form state and
+ * update the preview iframe if the form is valid.
+ *
+ * Dispatches the following events, in order:
+ * - `update` before sending the preview data to the server. Cancelable.
+ * - `json` after the preview data update request is completed. Note that the
+ *   data may not be valid and the iframe is not reloaded yet at this point.
+ *   Event `detail`:
+ *   - `data`: the response data that indicates whether the submitted data was
+ *     valid and whether the preview is available.
+ * - `error` when an error occurs while updating the preview data.
+ *   Event `detail`:
+ *   - `error`: the error object that was thrown.
+ * - `load` before reloading the preview iframe. Cancelable.
+ * - `loaded` after the preview iframe has been reloaded.
+ * - `ready` when the preview is ready for further updates – only fired on initial load.
+ * - `updated` after an update cycle is finished. This may or may not involve a
+ *   refresh of the iframe (e.g. in the case of an invalid form).
+ */
+export class PreviewController extends Controller<HTMLElement> {
```

Comment:
```suggestion
/**
 * Controls the preview panel component to submit the current form state and
 * update the preview iframe if the form is valid.
 *
 * Dispatches the following events in this order.
 *
 * @fires PreviewController#update - Before sending the preview data to the server. Cancelable.
 * @fires PreviewController#json - After the preview data update request is completed.
 * @fires PreviewController#error - When an error occurs while updating the preview data.
 * @fires PreviewController#load - Before reloading the preview iframe. Cancelable.
 * @fires PreviewController#loaded - After the preview iframe has been reloaded.
 * @fires PreviewController#ready - When the preview is ready for further updates – only fired on initial load.
 * @fires PreviewController#updated - After an update cycle is finished.
 *
 * @event PreviewController#update
 * @type {CustomEvent}
 * @property {boolean} cancelable - Is Cancelable
 * @property {string} name - `w-preview:update`
 *
 * @event PreviewController#json
 * @type {CustomEvent}
 * @property {Object} detail
 * @property {PreviewDataResponse} detail.data - The response data that indicates whether the submitted data was valid and whether the preview is available.
 * @property {string} name - `w-preview:json`
 *
 * @event PreviewController#error
 * @type {CustomEvent}
 * @property {Object} detail
 * @property {Error} detail.error - The error object that was thrown.
 * @property {string} name - `w-preview:error`
 *
 * @event PreviewController#load
 * @type {CustomEvent}
 * @property {boolean} cancelable - Is Cancelable
 * @property {string} name - `w-preview:load`
 *
 * @event PreviewController#loaded
 * @type {CustomEvent}
 * @property {string} name - `w-preview:loaded`
 *
 * @event PreviewController#ready
 * @type {CustomEvent}
 * @property {string} name - `w-preview:ready`
 *
 * @event PreviewController#updated
 * @type {CustomEvent}
 * @property {string} name - `w-preview:updated`
 */
export class PreviewController extends Controller<HTMLElement> {
```

Maybe we could leverage JSDoc for a bit of a nicer docs for these events 
https://jsdoc.app/tags-event
https://jsdoc.app/tags-fires

Reply:
TIL, although it doesn't seem to render nicely when hovering on the `PreviewController` class in VSCode. Is that expected?

Reply:
Yeah, it seems that VSCode does not yet support this. I still think it's a better way to articulate what you were trying to do.

I think that longer term we may want to consider generating autodocs from these somehow (for the events we want to officially support in the docs). Thanks for updating this.